### PR TITLE
Allow reuse of journal loading logic

### PR DESCRIPTION
**Pull request by chbatey**
Probably can be generalised even more as isn't specific to persistence any more. 

At present i've tested this with the akka persistence updater extension.

### CODE REVIEW

File: akka-persistence/src/main/scala/akka/persistence/Plugins.scala [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+  private[persistence] case class PluginHolder[ScalaDsl, JavaDsl](
+    scaladslPlugin: ScalaDsl, javadslPlugin: JavaDsl)
+    extends Extension
+}
+
+/**
+  * INTERNAL API
+  */
+@InternalApi
+private[akka] trait PluginProvider[T, ScalaDsl, JavaDsl] {
```
Comment (chbatey) [inline akka-persistence/src/main/scala/akka/persistence/Plugins.scala]: We use different names in various plugins. Went with a type class approach as not to change the public traits as java gets confused if we had a default impl of these methods that referes to the abstract method.

File: akka-persistence/src/main/scala/akka/persistence/Plugins.scala [hunk 2]
```diff
@@ -0,0 +1,99 @@
+/*
```
Comment (patriknw) [inline akka-persistence/src/main/scala/akka/persistence/Plugins.scala]: rename file to Plugin.scala

File: akka-persistence/src/main/scala/akka/persistence/Plugins.scala [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+import scala.annotation.tailrec
+import scala.reflect.ClassTag
+import scala.util.Failure
+
+/**
+  * INTERNAL API
+  */
+@InternalApi
+private[akka] object Plugin {
+  private[persistence] case class PluginHolder[ScalaDsl, JavaDsl](
```
Comment (patriknw) [inline akka-persistence/src/main/scala/akka/persistence/Plugins.scala]: final case class

File: akka-persistence-query/src/main/scala/akka/persistence/query/PersistenceQuery.scala [hunk 4]
```diff
@@ -24,23 +21,19 @@ object PersistenceQuery extends ExtensionId[PersistenceQuery] with ExtensionIdPr
...
 
-  def lookup() = PersistenceQuery
+  def lookup(): PersistenceQuery.type = PersistenceQuery
 
-  /** INTERNAL API. */
-  private[persistence] case class PluginHolder(
-    scaladslPlugin: scaladsl.ReadJournal, javadslPlugin: akka.persistence.query.javadsl.ReadJournal)
-    extends Extension
+  @InternalApi
+  private[akka] implicit val pluginProvider = new PluginProvider[ReadJournalProvider, scaladsl.ReadJournal, javadsl.ReadJournal] {
```
Comment (patriknw) [inline akka-persistence-query/src/main/scala/akka/persistence/query/PersistenceQuery.scala]: explicit type here?

File: akka-persistence-query/src/main/scala/akka/persistence/query/PersistenceQuery.scala [hunk 5]
```diff
@@ -24,23 +21,19 @@ object PersistenceQuery extends ExtensionId[PersistenceQuery] with ExtensionIdPr
...
 }
 
-class PersistenceQuery(system: ExtendedActorSystem) extends Extension {
-  import PersistenceQuery._
-
-  private val log = Logging(system, getClass)
-
-  /** Discovered query plugins. */
-  private val readJournalPluginExtensionIds = new AtomicReference[Map[String, ExtensionId[PluginHolder]]](Map.empty)
+import PersistenceQuery.pluginProvider
```
Comment (patriknw) [inline akka-persistence-query/src/main/scala/akka/persistence/query/PersistenceQuery.scala]: add comment that this is about the implicit

Comment (patriknw) [reply akka-persistence-query/src/main/scala/akka/persistence/query/PersistenceQuery.scala]: could we move it to some more specific place than at the top file level?

**Review by patriknw — APPROVED** [review]
LGTM, some nitpicks
Any mima filters?

File: akka-persistence-query/src/main/scala/akka/persistence/query/PersistenceQuery.scala (hunk 5 above)
Comment (chbatey) [reply akka-persistence-query/src/main/scala/akka/persistence/query/PersistenceQuery.scala]: I'll change this to pass in the implicit explicitly so nothing needs to be implicit

File: akka-persistence/src/main/scala/akka/persistence/Plugin.scala [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+
+import scala.annotation.tailrec
+import scala.reflect.ClassTag
+import scala.util.Failure
+
+/**
+ * INTERNAL API
+ */
+@InternalApi
+private[akka] object Plugin {
```
Comment (ktoso) [inline akka-persistence/src/main/scala/akka/persistence/Plugin.scala]: PersistencePlugin to avoid popping up so much in completion?

Comment (chbatey) [reply akka-persistence/src/main/scala/akka/persistence/Plugin.scala]: Yep. Will rename

**Review by ktoso — APPROVED** [review]
LGTM, nice to get reuse of this piece of code, has been replicated too many times :)

**Pull request merged**